### ISSUE TITLE
Source of non-determinism in the algorithms

### ISSUE THREAD

**Issue by louis-shao** [hill-a/stable-baselines#485]
I was trying the following code to test if the learning algorithms here are deterministic,
```bash
CUDA_VISIBLE_DEVICES= python -c "\
from stable_baselines import PPO2; \
PPO2('MlpPolicy', 'PongDeterministic-v4', verbose=1).learn(300, seed=10)"
```
The output on the screen prints the losses at step 128 and 256. They are not deterministic across different runs, even though I'm using CPU in this case. 

After some investigation, I believe the reason is because the call to `tf.set_random_seed` is too late in the library. According to the [documentation of `tf.set_random_seed`](https://www.tensorflow.org/api_docs/python/tf/random/set_random_seed), the function sets the graph level seed of the default graph, and if the graph level seed is not created (while creating an op), the op chooses an op-level seed randomly. In the stable baselines library, the set_random_seed function is called in the `model.learn` function, which is after the majority of the RL model graph is created. I tried adding a `set_global_seeds` right after the creation of the graph and the with statement in [this line](https://github.com/hill-a/stable-baselines/blob/0133b86413fc7d880092a29f2044ac890c555192/stable_baselines/ppo2/ppo2.py#L121) (of course we need some way to pass the seed from the user), the commandline in the beginning of the post suddenly becomes deterministic. The output is like the following. Except for the "fps" and "time_elapsed", the output is the same across different runs. 
```text
------------------------------------
| approxkl           | 0.00080277  |
| clipfrac           | 0.0         |
| explained_variance | -0.559      |
| fps                | 95          |
| n_updates          | 1           |
| policy_entropy     | 1.79106     |
| policy_loss        | -0.00494002 |
| serial_timesteps   | 128         |
| time_elapsed       | 2.62e-06    |
| total_timesteps    | 128         |
| value_loss         | 0.090288    |
------------------------------------
------------------------------------
| approxkl           | 0.000227978 |
| clipfrac           | 0.0         |
| explained_variance | 0           |
| fps                | 139         |
| n_updates          | 2           |
| policy_entropy     | 1.78901     |
| policy_loss        | 8.70687e-05 |
| serial_timesteps   | 256         |
| time_elapsed       | 1.34        |
| total_timesteps    | 256         |
| value_loss         | 0.113505    |
------------------------------------
```
I hope this can be fixed in all the algorithms.

**Comment by araffin** [comment]
Hello,
this issue is a duplicate of #145 and #285.
Contributions are welcomed to fix this issue ;)

**Comment by louis-shao** [comment]
> Hello,
> this issue is a duplicate of #145 and #285.
> Contributions are welcomed to fix this issue ;)

Hey, this is not quite a duplicate. I pointed out the problem and gave a simple fix. I just don't know how you guys want to structure the code so I'm leaving this to you guys.

**Comment by louis-shao** [comment]
The same problem I pointed out doesn't happen in the original baselines repo, because the `set_global_seeds` called [here](https://github.com/openai/baselines/blob/master/baselines/ppo2/ppo2.py#L80) is at the time before the construction of most of the graph pieces. That causes the creation of a default graph, and the random seed is set after that. Therefore the seed controls everything in the graph.

**Comment by araffin** [comment]
> this is not quite a duplicate.

Well, the issue is about determinism, which is a feature that we want and that is documented in #145 and #285.
If your simple fix works on all algorithms, then please submit a PR ;) (because currently the seeding does not work properly and this is an issue)
(Don't forget to read the contribution guide if you do so)

**Comment by louis-shao** [comment]
It is a simple fix, just call set_global_seeds after the with statement that is right after the creation of the graph. However, doing so would require passing a seed as an argument in the constructor of the learning algorithms. I'm not sure whether that would break the assumptions in the software architecture design. If the seed is only supposed to be passed in the learn function, after the construction of the graph, there is no solution here since tensorflow currently does not allow resetting the random state in the middle of training, but you can workaround that by having the random numbers generated from the numpy side, and plug into the tensorflow graphs, and put those generation in the learn function.

**Comment by araffin** [comment]
>If the seed is only supposed to be passed in the learn function,

Normally, we would avoid changing the API, but because this is currently not working, we can add the seed parameter to the constructor. I would also add a warning if users still pass a seed to the learn method.

**Comment by araffin** [comment]
@louis-shao I tried your fix, but this does not work for me (cf deterministic fix branch, tests/test_deterministic.py)...
In fact, calling `set_global_seeds` before creating the model should do the trick, no?

**Comment by araffin** [comment]
I tested your method for all algorithms (cf PR), it works for most of them, except TD3 for now.
